### ISSUE TITLE
[Improvement] Session number/character in OIA if x3270 is connected to IBM i

### ISSUE THREAD

**Issue by Dizzy-cz** [pmattes/x3270#91]
Hello, 
I tried using s3270 for automate tasks on IBM i, and this works pretty well.

I thinking about starting to use x3270 instead of official IBM i Terminal  Emulator (ACS), because I like some of other features of the x3270, like how much time system spent doing a task or history function.  I don't mind suppord only 24x80 characters, but I miss session identification in Operator Information Area.
It's possible to add this into OIA or is this information lost/not present if 3270 terminal is used?

Best Regards, Lukas

On screenshot: ACS with session B and E and c3270 without session identification.
![acs_vs_c3270](https://github.com/user-attachments/assets/9ce46590-948a-4d9b-ad40-695a5425aecd)

**Comment by pmattes** [comment]
Could you tell me a bit more about what 'session identification' is on a 5250 terminal? There is no such concept on a 3270, that I know of.

In TN3270/TN3270E, there is the notion of a Logical Unit, which c3270 does display, though it is on the right-hand side of the OIA rather than on the left. This is supplied by the host in either a TN3270E negotiation message or in an SNA BIND message. If the i Series host sends the 'session identification' to c3270 in this way, c3270 will display it.

**Comment by Dizzy-cz** [comment]
I'm still new to IBM i or z/OS so I don't know much.

However, there seems to me something called "Workstation ID" in the NT5250, probably similar to LU in the TN3270, but I don't know. 
I think "workstation ID" is set on terminal side and send to server as part of connection negotiation.
There is a WorkstationID variable in the IBM terminal Emulator (ACS).

I found this: https://www.ibm.com/support/pages/making-workstation-go-different-subsystem
If I understand this correctly, there must be some rule set on IBM i side to connect the terminal to correct subsystem.
Since there is no rule for LU(or other identification) that was send by x3270/c3270 all these terminals are connected to default subsystem QPADEV000S instead of specific subsystem for each connection. 

Or I'm complete wrong? 


So the "S" in OIA is Logical Unit (because it's connected to QPADEV000S subsystem) or something different?
![OIA](https://github.com/user-attachments/assets/00fa8352-ffe1-46db-9582-b8fe23fa9f25)

**Comment by Dizzy-cz** [comment]
I'm digging into it a bit, I think I understand some of the basics more.

This are my findings, I could be easily complete wrong:
The session identification on The IBM Terminal Emulator (A,B,C,D etc in left side) is only on Terminal side and does not send this information to IBM i sever or so. 
The terminal starts naming session from A to Z (and re-fill unused characters)  and send Workstation ID with  "=" at the end  (Collision Avoidance ID) so Workstation ID will gets suffix  from 0 to x (and re-fill unused numbers)
So you have "synced" Session A with workstation ID xxxx0, Session B with xxxx1 and so on. It's very useful in some cases.

I understand this can't be be done in same way in x3270/c3270/s3270, but what about displaying Workstation ID in OIA?
Is "Collision Avoidance ID" supported in the devname parameter?

**Comment by pmattes** [comment]
I can certainly look into displaying the RFC 4777 workstation ID in the OID, in the location usually used for the LU name, if no 3270 LU name is specified or returned by the host.

x3270 does not have support for collision avoidance IDs. But I can also look into implementing that, and displaying the one that finally worked in the OIA. If I understand correctly, if the workstation ID ends with one or more '=' characters, the emulator would replace those with a zero-padded number, starting with 0. If the host asks for the workstation ID again, the emulator would increment the number and send the new string. After all of the '=' characters were replaced by '9' characters, the emulator would give up and send all-nines again.

Does that sound reasonable?

The big issue here is time to implement it, but doesn't sound too difficult.

**Comment by Dizzy-cz** [comment]
Yes, I understand how "Collision Avoidance"  works in the same way, my test with IBM terminal Emulator and IBM i server supports this understanding. 

Your suggested implementation sounds great, even better that implementation the IBM Terminal Emulator. :)

I totally get it, It's just "nice to have" feature, nothing extra important. :)
I'm glad you took the time to even think about it.
